### ISSUE TITLE
there are no tests for thread-safety

### ISSUE THREAD

**Issue by yegor256** [yegor256/factbase#98]
We claim that `Factbase` class is thread-safe, but there are no tests that confirm this. Let's use [`threads`](https://github.com/yegor256/threads) Gem, as explained [here](https://www.yegor256.com/2018/11/06/ruby-threads.html), to create such tests.

**Comment by Suban05** [comment]
@yegor256 do I understand correctly, that this class is  really thread-safe, but we don't have any tests to approve this fact?

It's just that if I open the code of the class, then I see this comment:
https://github.com/yegor256/factbase/blob/2e2dae296e2273aefee245240c82865122780df2/lib/factbase.rb#L77

**Comment by yegor256** [comment]
@Suban05 I don't think it's thread safe. If you do `fb.insert.foo = 42` from 100 different threads at the same time, you won't see 100 facts in `fb`. Most probably, there will be fewer.

**Comment by Suban05** [comment]
@yegor256 if I understood your described case correctly, this test passes:
```rb
def test_concurrent_inserts
  fb = Factbase.new
  Threads.new(100).assert do
    fb.insert.foo = 42
  end
  assert_equal(100, fb.size)
end
```
that's very good

**Comment by Suban05** [comment]
@yegor256 of course, I will do it
assign me, please)

**Comment by Suban05** [comment]
@yegor256 this test fails:
```rb
def test_concurrent_queries
  fb = Factbase.new
  Threads.new(100).assert do |i|
    fact = fb.insert
    fact.thread_id = i
    fact.value = i * 10
  end
  Threads.new(100).assert do
    results = fb.query('(exists thread_id)').each.to_a
    assert_equal(100, results.size)

    thread_ids = results.map { |fact| fact.thread_id }
    assert_equal((0..99).to_a, thread_ids.sort)
  end
end
```
When I execute a query, all the facts have `thread_id = 0`.

But this test is OK:
```rb
def test_concurrent_queries
  fb = Factbase.new
  100.times do |i|
    fact = fb.insert
    fact.thread_id = i
    fact.value = i * 10
  end
  Threads.new(100) do
    results = fb.query('(exists thread_id)').each.to_a
    assert_equal(100, results.size)

    thread_ids = results.map { |fact| fact.thread_id }
    assert_equal((0..99).to_a, thread_ids.sort)
  end
end
```

Do I do something incorrectly or is it a mistake?

**Comment by yegor256** [comment]
@Suban05 looks weird. I don't know why the test fails. You wrote it correctly.

**Comment by Suban05** [comment]
@yegor256 sometimes this test also fails:

```rb

def test_concurrent_transactions_successful
  fb = Factbase.new
  Threads.new(100).assert do |i|
    fb.txn do |fbt|
      fact = fbt.insert
      fact.thread_id = i
      fact.value = i * 10
    end
  end
  facts = fb.query('(exists thread_id)').each.to_a
  assert_equal(100, facts.size)
  facts.each do |fact|
    assert_equal(fact.value, fact.thread_id * 10)
  end
end
```
it's very rare, but I caught the situation, when `facts.size` equalled `95`, instead of `100`

What will we do? We can't solve this problem, because we really are not thread-safe

**Comment by yegor256** [comment]
@Suban05 we must be thread-safe. You can add a test to the code base, but `skip` it, making a puzzle inside (read about [PDD](https://www.yegor256.com/2010/03/04/pdd.html)). Thus, you will earn some points for the PR and we'll continue from the puzzle.

**Comment by yegor256** [comment]
@Suban05 this code is NOT thread-safe:

```
arr = []
Threads.new(5).assert do |i|
  arr << i
end
```

All five threads are trying to append a number to the array, but only one thread "wins" and adds the number, for example `0`.

**Comment by 0pdd** [comment]
@yegor256 3 puzzles [#115](https://github.com/yegor256/factbase/issues/115), [#116](https://github.com/yegor256/factbase/issues/116), [#117](https://github.com/yegor256/factbase/issues/117) are still not solved.

**Comment by 0pdd** [comment]
@yegor256 all 3 puzzles are solved here: [#115](https://github.com/yegor256/factbase/issues/115), [#116](https://github.com/yegor256/factbase/issues/116), [#117](https://github.com/yegor256/factbase/issues/117).
